agreement/events.go

```diff
@@ -975,50 +1006,59 @@ func (e checkpointEvent) AttachConsensusVersion(v ConsensusVersionView) external
 	return e
 }
 
+// This timestamp is assigned to messages that arrive for round R+1 while the current player
+// is still waiting for quorum on R.
+const pipelinedMessageTimestamp = time.Nanosecond
+
+type constantRoundZeroTimer time.Duration
+
+func (c constantRoundZeroTimer) Since() time.Duration { return time.Duration(c) }
+
+// clockForRound retrieves the roundZeroTimer used for AttachValidatedAt and AttachReceivedAt.
+func clockForRound(currentRound round, currentClock roundZeroTimer, historicalClocks map[round]roundZeroTimer) func(round) roundZeroTimer {
+	return func(eventRound round) roundZeroTimer {
+		if eventRound > currentRound {
+			return constantRoundZeroTimer(pipelinedMessageTimestamp)
+		}
+		if eventRound == currentRound {
+			return currentClock
+		}
+		if clock, ok := historicalClocks[eventRound]; ok {
+			return clock
+		}
+		return constantRoundZeroTimer(0)
+	}
+}
+
 // AttachValidatedAt looks for a validated proposal or vote inside a
 // payloadVerified or voteVerified messageEvent, and attaches the given time to
 // the proposal's validatedAt field.
-func (e messageEvent) AttachValidatedAt(d time.Duration, currentRound round) messageEvent {
+func (e messageEvent) AttachValidatedAt(getClock func(eventRound round) roundZeroTimer) messageEvent {
 	switch e.T {
 	case payloadVerified:
-		if e.Input.Proposal.Round() > currentRound {
-			e.Input.Proposal.validatedAt = 1
-		} else {
-			e.Input.Proposal.validatedAt = d
-		}
+		e.Input.Proposal.validatedAt = getClock(e.Input.Proposal.Round()).Since()
 	case voteVerified:
-		if e.Input.Vote.R.Round > currentRound {
-			e.Input.Vote.validatedAt = 1
-		} else {
-			e.Input.Vote.validatedAt = d
-		}
+		e.Input.Vote.validatedAt = getClock(e.Input.Vote.R.Round).Since()
 	}
 	return e
 }
 
 // AttachReceivedAt looks for an unauthenticatedProposal inside a
 // payloadPresent or votePresent messageEvent, and attaches the given
 // time to the proposal's receivedAt field.
-func (e messageEvent) AttachReceivedAt(d time.Duration, currentRound round) messageEvent {
-	if e.T == payloadPresent {
-		if e.Input.UnauthenticatedProposal.Round() > currentRound {
-			e.Input.UnauthenticatedProposal.receivedAt = 1
-		} else {
-			e.Input.UnauthenticatedProposal.receivedAt = d
-		}
-	} else if e.T == votePresent {
+func (e messageEvent) AttachReceivedAt(getClock func(eventRound round) roundZeroTimer) messageEvent {
```

Comment:
For a little while I was asking myself "Oh, this modifies the the messageEvent, that's weird, since we don't normally do that."

Is there any naming convention that would help. I sometimes use "WithRecievedAt" to make it clear I'm making a new object with a field assigned.

Low priority. I should have realized that agreement code is always by value.

Reply:
Ugh, nevermind, `messageEvent` is a struct, not an interface.